tempesta_fw/http_limits.c

```diff
@@ -715,266 +671,319 @@ enum {
+		goto block;
+	}
+
+	return frang_http_req_hdrs_check(ra, conn, data);
+block:
+	return TFW_BLOCK;
+}
+
+static int
+frang_http_req_incomplete_body_check(FrangAcc *ra, TfwConn *conn,
+				     TfwFsmData *data)
+{
+	TfwHttpReq *req = (TfwHttpReq *)data->req;
+	__FRANG_CFG_VAR(body_len, http_body_len);
+	__FRANG_CFG_VAR(body_timeout, clnt_body_timeout);
+	__FRANG_CFG_VAR(bchunk_cnt, http_bchunk_cnt);
+	struct sk_buff *skb = data->skb;
+
+	frang_dbg("check incomplete request body for client %s, acc=%p\n",
+		  &FRANG_ACC2CLI(ra)->addr, ra);
+
+	/* CLRF after headers was parsed, but the body didn't arrive yet. */
+	if (TFW_STR_EMPTY(&req->body))
+		return TFW_PASS;
+	/*
+	 * Ensure that HTTP request body is coming without delays.
+	 * The timeout is between chunks of the body, so reset
+	 * the start time after each check.
+	 */
+	if (body_timeout && req->tm_bchunk && (skb != req->body.skb)) {
+		unsigned long start = req->tm_bchunk;
+		unsigned long delta = body_timeout;
+
+		if (time_is_before_jiffies(start + delta)) {
+			frang_limmsg("client body timeout",
+				     jiffies_to_msecs(jiffies - start),
+				     jiffies_to_msecs(delta),
+				     &FRANG_ACC2CLI(ra)->addr);
+			goto block;
+		}
+	}
+
+	/* Limit number of chunks in request body */
+	if (bchunk_cnt && (req->chunk_cnt > bchunk_cnt)) {
+		frang_limmsg("HTTP body chunk count", req->chunk_cnt,
+			     bchunk_cnt, &FRANG_ACC2CLI(ra)->addr);
+		goto block;
+	}
+
+	if (body_len && (req->body.len > body_len)) {
+		frang_limmsg("HTTP body length", req->body.len,
+			     body_len, &FRANG_ACC2CLI(ra)->addr);
+		goto block;
+	}
+
+
+	return TFW_PASS;
+block:
+	return TFW_BLOCK;
+}
+
+static int
+frang_http_req_trailer_check(FrangAcc *ra, TfwConn *conn,
+			     TfwFsmData *data)
+{
+	int r = TFW_PASS;
+	TfwHttpMsg *hmreq = (TfwHttpMsg *)data->req;
+	const TfwStr *field, *end, *dup, *dup_end;
+
+	if (!test_bit(TFW_HTTP_B_CHUNKED_TRAILER, hmreq->flags))
+		return TFW_PASS;
+	/*
+	 * Don't use special settings for the trailer part, keep on
+	 * using body limits.
+	 */
+	r = frang_http_req_incomplete_body_check(ra, conn, data);
+	if (!r)
+		r = frang_http_req_hdrs_check(ra, conn, data);
+	if (r)
+		return r;
+
+	if (!tfw_http_parse_is_done(hmreq))
+		return TFW_POSTPONE;
+	/*
+	 * Block request if the same header appear in both main and
+	 * trailer headers part. Some intermediates doesn't read trailers, so
+	 * request processing may differ.
+	 */
+	FOR_EACH_HDR_FIELD(field, end, hmreq) {
+		int trailers = 0, dups = 0;
+		TFW_STR_FOR_EACH_DUP(dup, field, dup_end) {
+			trailers += !!(dup->flags & TFW_STR_TRAILER);
+			dups += 1;
+		}
+		if (trailers && (dups != trailers)) {
+			frang_msg("HTTP field appear in header and trailer "
+				  "for client %p",
+				  &FRANG_ACC2CLI(ra)->addr, "\n");
 			return TFW_BLOCK;
 		}
 	}
 
-	T_FSM_START(req->frang_st) {
+	return r;
+}
+
+static int
+frang_http_req_process(FrangAcc *ra, TfwConn *conn, TfwFsmData *data)
```

Comment:
How the patch fixes the problem mentioned in https://github.com/tempesta-tech/tempesta/issues/1028 ?

> The other related issue - per-vhost limits. Imagine request formed by two skbs: headers and body. Target vhost is not known until request is parsed up to end. Frang is called twice for this request: at EOS and at message end. The headers are checked agains defaut frang limits. Only body is checked against vhost frang limit. Again root of the issue is frang FSM.

Currently Frang uses `__FRANG_CFG_VAR`: if we have vhost location, then we use limits from it, otherwise we use global default limits. The problem is that `tfw_location_match()` is called by`tfw_http_req_process()` when we fully process a request instead of be called when we have URI host aprt and/or Host header. So `TFW_HTTP_FSM_REQ_CHUNK` is always called with global limits.

This problem seems can be fixed by calling the location logic from HTTP parser. This location should be used for limiting. Later HTTPtables may assign a new location and/or vhost, so we need to reassign `req->location` and `req->vhost` if they are changed. Next, if vhost/location sets more strict limits than the global settings, we have to check URI, Host and other limits, which were passed before getting the location/vhost, against the vhost/location settings. This implies relatively complex processing.

So I propose to always use global limits on `TFW_HTTP_FSM_REQ_CHUNK` hook and on `TFW_HTTP_FSM_REQ_MSG` check all the limits again.

Please create a new chapter in https://github.com/tempesta-tech/tempesta/wiki/HTTP-security about per-location/vhost logic. Move 

> Besides "frang_limits" section, security limits can be specified inside location directives to apply them for particular URLs and for particular virtual hosts (for details see Virtual hosts and locations and Local HTTP security limits).

to the chapter and extend the chapter with **example** of the per-vhost and per-location limiting exmplaing that URI. headers and body are checked against the default limit first and then by a perf-vhost/location limit. There should be explanation how the limiting machinery works and that per-vhost/location limiting is expensive, so a use should avoid double limitation on global and per-vhost/location layer if they need more performance, but assembling the whole message w/o limits is DoS dangerous, so per-vhost/location limits must be used with caution.

Reply:
> per-vhost limits

Yes, that wasn't fixed. Things got more complicated since the issue was first reported and I think we should discuss the better solution and implement it in a new PR. Now we have more precise client assignment for the incoming requests, see #1115, every new request may belong not only to a new vhost, but it can also have a new set of accounting data. Thus we can't normally switch between `TFW_HTTP_FSM_REQ_CHUNK` and `TFW_HTTP_FSM_REQ_MSG` hooks (accounting data may differ between them).

The main problem is to make frang limits **predictable** and understandable for end user.

> This problem seems can be fixed by calling the location logic from HTTP parser.

It seems like only URI and `Host:` header are required to set correct location, but that's not true. The appropriate vhost must be found first. Since HTTP tables are complicated, vhost can't be identified without full headers set. This logic mustn't be in the parser. But it can be in `tfw_http_req_process()`: vhost and location may be updated before  calling `TFW_HTTP_FSM_REQ_CHUNK` hook. I've done something pretty similar when i was working on #498.

But it's not a solution. Global frang limits set and per-vhost limits mostly always differ, and decision whether a request is blocked or not depends on how the request is split between skbs. It's absolutely unpredictable behaviour.

> So I propose to always use global limits on TFW_HTTP_FSM_REQ_CHUNK hook and on TFW_HTTP_FSM_REQ_MSG check all the limits again.

That is also not good solution. First it's ineffective, secondly some checks can be done twice, e.g. `client_body_timeout`.

The exact solution must be more complex. First, let's assume, that target location and vhost are updated right after all the headers are parsed and right before `TFW_HTTP_FSM_REQ_CHUNK` hook. Now if we look at both global frang limits and per-vhost/location limits, we will see some relations between them:

| Limit | Where can be defined | Relation between Global and Per-Vhost |
| --- | --- | --- |
| ip_block | global only |  -  |
| request_rate | Both | Per-Vhost must be more strict or equal |
| request_burst | Both | Per-Vhost must be more strict or equal |
| connection_rate | global only |  -  |
| connection_burst | global only |  -  |
| concurrent_connections | global only |  -  |
| client_header_timeout | global only |  |
| client_body_timeout | Both | Global is ignored, per-vhost is always used |
| http_uri_len | Both | Per-Vhost must be more strict or equal |
| http_field_len | Both | Per-Vhost must be more strict or equal |
| http_body_len | Both | Global is ignored, per-vhost is always used |
| http_header_cnt | Both | Per-Vhost must be more strict or equal |
| http_header_chunk_cnt | Both | Per-Vhost must be more strict or equal |
| http_body_chunk_cnt | Both | Global is ignored, per-vhost is always used |
| http_host_required | Both | Global is ignored, per-vhost is always used |
| http_ct_required | Both | Global is ignored, per-vhost is always used |
| http_methods | Both | Per-Vhost must be more strict or equal |
| http_ct_vals | Both | Global is ignored, per-vhost is always used |
| http_resp_code_block | Both | Global is ignored, per-vhost is always used |

It's clear, that we need to define different directives sets for global and per-vhost limits. 

Some configuration directives (like `srv_group`) has inheritance of settings. When talking about frang, we have multiple levels: global - vhost - location. It's becomes more and more complicated to create a lot of vhosts with non-default frang settings. I have a suggestion upon that:
```
frang_global_policy {
 <global-only + 'both (global|per-vhost)' directives>
}

frang_policy "my_policy" {
 <only per-vhost directives>
}

frang_policy "my_policy_location" {
 <only per-vhost directives>
}

vhost my_vhost {
    frang_policy "my_policy";
}

vhost my_vhost_2 {
    frang_policy "global";
}
```

Reply:
Frang limits are simple limitation and the per-vhost/location logic must not be complex. I agree that double checking isn't fast and that's bad. If we can implement this in a way that if we have only global or per-vhost/location limit then we do not lose performance (check limits only once) and slow down only if both of the limit layers are specified, then that would be fine.

Reply:
It seems now we have single pass over the limit and that's fine.